### PR TITLE
don't return Service Unavailable status when query is rejected

### PR DESCRIPTION
**Pull request by alexkorotkikh**
instead just don't return any datapoints

### CODE REVIEW

**Comment by vetinari** [comment]
:+1:

File: src/main/java/org/kairosdb/core/http/rest/json/QueryParser.java [hunk 1]
```diff
@@ -237,6 +237,8 @@ private void validateMetricTier(final Metric metric) throws QueryRejectedExcepti
 				ret.add(queryMetric);
 			} catch (ContextualJsonSyntaxException e) {
 				throw new BeanValidationException(new SimpleConstraintViolation(e.getContext(), e.getMessage()), context);
+			} catch (QueryRejectedException e) {
```
Comment (a1exsh) [inline src/main/java/org/kairosdb/core/http/rest/json/QueryParser.java]: Since we are suppressing the exception here and continue iterating over the `metricsArray`, I wonder if the resulting gap is not going to break rendering?

Consider a query of 3 metrics where first and last are not rejected, but the one in the middle of the list is.  We then calculate `context` as `query.metric[0]` and `query.metric[2]`, but we skip over `query.metric[1]`.
I'm not all that familiar with the KDB API, but I think we should verify what happens on the client (i.e. Grafana) once it gets fewer items than it expects in response.

I would prefer that we return the same number of items in response as we would if rejection was not activated (requested 3 metrics, returned 3 metrics), simply with the empty list of data points for the rejected ones.

Comment (pitr) [reply src/main/java/org/kairosdb/core/http/rest/json/QueryParser.java]: agreed! after quick discussion, this is what we need to - return empty list of datapoints

File: src/main/java/org/kairosdb/core/http/rest/MetricsResource.java [hunk 2]
```diff
@@ -233,7 +232,7 @@ public Response addGzip(@Context HttpHeaders httpHeaders, InputStream gzip) {
 	public Response add(@Context HttpHeaders httpHeaders, InputStream json) {
 		Span span = createSpan("datapoints_insert", httpHeaders);
 
-		try (Scope scope = tracer.scopeManager().activate(span, false)) {
```
Comment (a1exsh) [inline src/main/java/org/kairosdb/core/http/rest/MetricsResource.java:236]: This is unrelated to the actual change, right?

File: src/main/java/org/kairosdb/core/http/rest/MetricsResource.java [hunk 3]
```diff
@@ -438,7 +437,12 @@ public Response get(@Context HttpHeaders httpHeaders, String json) throws Except
...
 						DatastoreQuery dq = datastore.createQuery(query);
 						try {
-							List<DataPointGroup> results = dq.execute();
+							List<DataPointGroup> results;
+							if (query.isRejected()) {
+								results = new ArrayList<>();
+							} else {
+								results = dq.execute();
+							}
 							jsonResponse.formatQuery(results, query.isExcludeTags(), dq.getSampleSize());
```
Comment (a1exsh) [inline src/main/java/org/kairosdb/core/http/rest/MetricsResource.java]: I wonder how `getSampleSize()` is going to behave if we didn't call `execute()`?

File: src/main/java/org/kairosdb/core/http/rest/MetricsResource.java [hunk 4]
```diff
@@ -406,16 +406,15 @@ public Response get(@Context HttpHeaders httpHeaders, String json) throws Except
...
 			jsonResponse.begin();
 			List<QueryMetric> queries = queryParser.parseQueryMetric(json);
 
-
 			return limiter.callWithTimeout(() -> {
-				try (Scope internalScope = tracer.scopeManager().activate(span, false)) {
+				try (Scope internalScope = tracer.scopeManager().activate(span)) {
 
 					for (QueryMetric query : queries) {
 						Map<String, Collection<String>> tags = query.getTags().asMap();
```
Comment (a1exsh) [inline src/main/java/org/kairosdb/core/http/rest/MetricsResource.java:421]: Does the call to `getTags()` result in accessing the data store?  Or do we get an empty map in case query is marked for rejection?

File: src/main/java/org/kairosdb/core/http/rest/MetricsResource.java (hunk 2 above)
Comment (alexkorotkikh) [reply src/main/java/org/kairosdb/core/http/rest/MetricsResource.java:236]: yep, just a little clean up

File: src/main/java/org/kairosdb/core/http/rest/MetricsResource.java (hunk 3 above)
Comment (alexkorotkikh) [reply src/main/java/org/kairosdb/core/http/rest/MetricsResource.java]: it's going to stay uninitialized, and the default value of `int` is 0

File: src/main/java/org/kairosdb/core/http/rest/MetricsResource.java [hunk 5]
```diff
@@ -438,7 +438,13 @@ public Response get(@Context HttpHeaders httpHeaders, String json) throws Except
 
 						DatastoreQuery dq = datastore.createQuery(query);
```
Comment (a1exsh) [inline src/main/java/org/kairosdb/core/http/rest/MetricsResource.java]: What's involved in creating an instance of this class?  Can/should we avoid doing that if we know the query is rejected?

File: src/main/java/org/kairosdb/core/http/rest/MetricsResource.java (hunk 4 above)
Comment (alexkorotkikh) [reply src/main/java/org/kairosdb/core/http/rest/MetricsResource.java:421]: this refers to the tags that were passed in a query by the user... so it doesn't require any access to data store

**Comment by a1exsh** [comment]
:+1:

**Comment by alexkorotkikh** [comment]
Tested on staging with service level config:
```yaml
id: zmon-service-level-config
type: zmon_config
team: ZMON
data:
  ingest_max_check_tier: 0
  query_distance_hours_limit: 0
  query_max_check_tier: 2
```
and check-tiers:
```yaml
id: zmon-check-tiers
type: zmon_config
team: ZMON
data:
  critical:
  - 9
  important:
  - 4
```
Request:
```json
{
  "cache_time": 0,
  "start_absolute": 1572923198000,
  "end_absolute": 1572944798000,
  "metrics": [
    {
      "limit": 0,
      "name": "zmon.check.9",
      "tags": {},
      "aggregators": [],
      "group_by": []
    },
    {
      "limit": 0,
      "name": "zmon.check.8",
      "tags": { "application": ["cdp"] },
      "aggregators": [],
      "group_by": []
    },
    {
      "limit": 0,
      "name": "zmon.check.4",
      "tags": {},
      "aggregators": [],
      "group_by": []
    }
  ]
}
```
Result:
image

**Comment by alexkorotkikh** [comment]
👍

File: logback.xml [hunk 6]
```diff
@@ -26,8 +26,5 @@
         <logger name="org.kairosdb.core.datastore.KairosDatastore" level="INFO" additivity="false">
                 <appender-ref ref="stdout"/>
         </logger>
-        <logger name="org.kairosdb.datastore.cassandra.CassandraClientImpl" level="DEBUG" additivity="false">
```
Comment (pitr) [inline logback.xml:29]: why was this removed?

Comment (alexkorotkikh) [reply logback.xml:29]: This was added some months ago to track cassandra client pool status, but appeared to be useless

**Comment by pitr** [comment]
:+1:
